### DISCUSSION TITLE
Integration best practices: pam/nss vs. OIDC

### DISCUSSION THREAD

**Discussion by TheRealGramdalf** [kanidm/kanidm#2361] [General]
In https://github.com/kanidm/kanidm/issues/2320#issuecomment-1819977443, @Firstyear  said:
> You should always use the pam/nss tools from Kanidm unixd if possible, over an other integration.

I'm curious about this - if you don't mind elaborating, take Proxmox for an example:
In proxmox, it is possible to log in to the WebUI via PAM/NSS or OIDC. Both provide a similar end result - in order to access the WebUI, you need to authenticate yourself through one of these "Realms". Once authenticated, Proxmox handles the rest (i.e. permissions specific to Proxmox).

In such a situation, which one would be preferable? Here's a couple of my thoughts:
- At least at the moment, kanidm is quite fluid. While I expect it to simmer down after the initial release (in terms of breaking changes), it means keeping a close eye on `kanidm-unixd` upgrades
- OIDC has a smaller scope - i.e. just the WebUI - unixd covers many more things, which greatly increases attack surface.

This is by no means a well thought out post, just a passing thought I had and was curious about.
Cheers!

**Comment by yaleman** [comment]
Proxmox is a special case in that it supports PAM auth on a Web UI 😄 I use the PAM module on it fine, OIDC's probably... better? 🤷🏻 

In context, the original comment was in reference to authentication to a _system_, rather than an _application_.

**Comment by Firstyear** [comment]
> In [#2320 (comment)](https://github.com/kanidm/kanidm/issues/2320#issuecomment-1819977443), @Firstyear said:
> 
> > You should always use the pam/nss tools from Kanidm unixd if possible, over an other integration.
> 
> I'm curious about this - if you don't mind elaborating, take Proxmox for an example: In proxmox, it is possible to log in to the WebUI via PAM/NSS or OIDC.

As james mentioned this comment is *specifically* about system authentication. 

The context of the comment matters here. The question was "Can we use SSSD with Kanidm". In that context, the answer is "No, SSSD is problematic for a host of reasons". 

It is not a comment saying "use pam for absolutely everything because it's the best". In fact, it's one of the *weakest* parts of the Kanidm authentication story, because linux and unix machines are so embedded in "username/password" and "text inputs" for authentication, that we simply can't consistently do better things via pam. That's why we push that you use ssh keys for remote access and limit the pam username/password to "sudo only", because that way you have a form of loose MFA protected by an initial cryptographic authentication. 

> Both provide a similar end result - in order to access the WebUI, you need to authenticate yourself through one of these "Realms". Once authenticated, Proxmox handles the rest (i.e. permissions specific to Proxmox).
> 
> In such a situation, which one would be preferable? Here's a couple of my thoughts:

Use OIDC. This is exactly why we offer it. 

> 
>     * At least at the moment, kanidm is quite fluid. While I expect it to simmer down after the initial release (in terms of breaking changes), it means keeping a close eye on `kanidm-unixd` upgrades
> 
>     * OIDC has a smaller scope - i.e. just the WebUI - unixd covers many more things, which greatly increases attack surface.

Unixd and OIDC both have *different* attack surfaces, both of which are *highly* audited to ensure we offer secure implementations for both. But those implementations are only as good as how they are used, and using unixd for proxmox is the wrong tool for the job.

**Comment by TheRealGramdalf** [comment]
Many thanks for the thorough explanation! I'm not super familiar with SSSD, so that's probably where the loss of context happened.
